include/LightGBM/category_encoding_provider.hpp

```diff
@@ -0,0 +1,771 @@
+  data_size_t num_data_;
+  // list of categorical feature indices (real index, not inner index of Dataset)
+  std::vector<int> categorical_features_;
+
+  // maps training data index to fold index
+  std::vector<int> training_data_fold_id_;
+  // prior used by per fold
+  std::vector<double> fold_prior_;
+  // weight of the prior in category encoding calculation
+  double prior_weight_;
+  // record whether a feature is categorical in the original data
+  std::vector<bool> is_categorical_feature_;
+
+  // number of features in the original dataset, without adding count features
+  int num_original_features_;
+  // number of features after converting categorical features
+  int num_total_features_;
+
+  // number of threads used for category encoding
+  int num_threads_;
+
+  // the accumulated count information for category encoding
+  std::unordered_map<int, std::vector<std::unordered_map<int, int>>> count_info_;
+  // the accumulated label sum information for category encoding
+  std::unordered_map<int, std::vector<std::unordered_map<int, double>>> label_info_;
+  // the accumulated count information for category encoding per thread
+  std::vector<std::unordered_map<int, std::vector<std::unordered_map<int, int>>>> thread_count_info_;
+  // the accumulated label sum information for category encoding per thread
+  std::vector<std::unordered_map<int, std::vector<std::unordered_map<int, double>>>> thread_label_info_;
+  // the accumulated label sum per fold
+  std::vector<double> fold_label_sum_;
+  // the accumulated label sum per thread per fold
+  std::vector<std::vector<double>> thread_fold_label_sum_;
+  // the accumulated number of data per fold per thread
+  std::vector<std::vector<data_size_t>> thread_fold_num_data_;
+  // number of data per fold
+  std::vector<data_size_t> fold_num_data_;
+  // categorical value converters
+  std::vector<std::unique_ptr<CatConverter>> category_encoders_;
+  // whether the old categorical handling method is used
+  bool keep_raw_cat_method_;
+
+  // temporary parser used when accumulating statistics from file
+  std::unique_ptr<Parser> tmp_parser_;
+  // temporary oneline_features used when accumulating statistics from file
+  std::vector<std::pair<int, double>> tmp_oneline_features_;
+  // temporary random generator used when accumulating statistics from file,
+  // used to generate training data folds for category encoding calculations
+  std::mt19937 tmp_mt_generator_;
+  // temporary fold distribution probability when accumulating statistics from file
+  std::vector<double> tmp_fold_probs_;
+  // temporary fold distribution when accumulating statistics from file
+  std::discrete_distribution<int> tmp_fold_distribution_;
+  // temporary feature read mask when accumulating statistics from files
+  std::vector<bool> tmp_is_feature_processed_;
+  // mark whether the category encoding statistics is accumulated from file
+  bool accumulated_from_file_;
+};
+
+class CategoryEncodingParser : public Parser {
```

Comment:
Use parser as middleware looks like wield, it is more like a step in pipeline processing. Not sure is that possible we refactor the processing steps to pipeline pattern?